### DISCUSSION TITLE
Using the unreleased rocket_db_pools with other Rocket-dependent crates

### DISCUSSION THREAD

**Discussion by benny-n** [rwf2/Rocket#1901] [Questions] [answered]
Hi, 
I'm developing a Rocket web application server which communicates with a mongo cluster via rust mongodb. To integrate the mongodb driver into Rocket I am using the unreleased crate "rocket_db_pools" which to my understanding does just that.
Furthermore, as a part of the authorization process, a user is required to login with google or github.
For the implementation of said authorization process I am using rocket_oauth2 version 0.5.0-rc.1.
The problem: 
rocket_oauth2 depends on Rocket version 0.5.0-rc.1. in its Cargo.toml file, but rocket_db_pools is unreleased, so I have to specify an explicit git dependency for it to work. 
Since one can't mix git dependencies with crates.io dependencies (as I recently learned after reading the Rocket docs FAQ) I can't use both at the same time. 
It seems to me that as long as rocket_db_pools is unreleased, using it will be incompatible with any crate that strictly depends on Rocket version 0.5.0-rc.1. (correct me if I'm wrong here).
So should I just wait patiently until rocket_db_pools is officially released? Or maybe there's a workaround that I'm not familiar with to sort these dependencies?
I would be grateful to hear your advices.
Thanks in advance!

**Comment by jebrosen** [comment, marked as answer]
> Or maybe there's a workaround that I'm not familiar with to sort these dependencies?

There is, somewhat; unfortunately it's a bit inconvenient to maintain. In particular, you could try using the [`[patch]` section in Cargo.toml](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html), and/or private forks and [`path` dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-path-dependencies).

In this case, you may be able to make a copy of `rocket_oauth2` that depends on `rocket` via git (at the same branch/tag/rev as everything else uses) instead of by version, and then either depend on the copy via `path` or by using `[patch]`. This approach can get quite complicated as you add more dependencies, of course, and it does add a bit more work to the updating and maintenance process. But, it can be quite useful for trying out prerelease versions or code changes in dependencies in the context of your whole project.

**Reply by benny-n** [reply]
Thanks for the answer, using ```[patch]``` is actually what I ended up doing to make it work.
